pure_protobuf/protobuf.py

```diff
@@ -1,211 +1,247 @@
-        '''
-        raise TypeError('Don\'t call this directly.')
-    
+        """
+        raise TypeError("Don't call this directly.")
+
     def dumps(self, value):
-        '''
+        """
         Dumps its value to string and returns this string.
-        '''
-        fp = cStringIO.StringIO()
+        """
+        fp = io.BytesIO()
         self.dump(fp, value)
         return fp.getvalue()
-    
+
     def loads(self, s):
-        '''
+        """
         Loads its value from a string and returns a read value.
-        '''
-        return self.load(cStringIO.StringIO(s))
-        
+        """
+        return self.load(io.BytesIO(s))
+
     def __hash__(self):
-        '''
+        """
         Returns a hash of this type.
-        '''
+        """
         return hash(self.__class__.__name__)
-        
+
+
 class UVarintType(Type):
-    '''
+    """
     Represents an unsigned Varint type.
-    '''
+    """
 
     WIRE_TYPE = 0
 
     def dump(self, fp, value):
         shifted_value = True
         while shifted_value:
             shifted_value = value >> 7
-            fp.write(chr((value & 0x7F) | (0x80 if shifted_value != 0 else 0x00)))
+            data = (value & 0x7F) | (0x80 if shifted_value != 0 else 0x00)
+            write_data = six.int2byte(data)
+            fp.write(write_data)
             value = shifted_value
-        
+
     def load(self, fp):
         value, shift, quantum = 0, 0, 0x80
         while (quantum & 0x80) == 0x80:
             quantum = ord(fp.read(1))
             value, shift = value + ((quantum & 0x7F) << shift), shift + 7
         return value
-        
+
+
 class VarintType(UVarintType):
-    '''
+    """
     Represents a signed Varint type. Implements ZigZag encoding.
-    '''
-    
+    """
+
     def dump(self, fp, value):
         encoded_varint = abs(value) << 1
         if value < 0:
             encoded_varint -= 1
         UVarintType.dump(self, fp, encoded_varint)
-        
+
     def load(self, fp):
         encoded_varint = UVarintType.load(self, fp) + 1
         div = encoded_varint >> 1
         return div if encoded_varint & 1 else -div
-      
+
+
 class BoolType(UVarintType):
-    '''
-    Represents a boolean type. Encodes True as UVarint 1, and False as UVarint 0.
-    '''
+    """
+    Represents a boolean type.
+    Encodes True as UVarint 1, and False as UVarint 0.
+    """
+
+    dump = lambda self, fp, value: fp.write(
+        b"\x01" if value else b"\x00"
+    )  # Similarly to UVarint.
 
-    dump = lambda self, fp, value: fp.write('\x01' if value else '\x00') # Similarly to UVarint.
-    
     load = lambda self, fp: UVarintType.load(self, fp) != 0
-        
+
+
 class BytesType(Type):
-    '''
+    """
     Represents a raw bytes type.
-    '''
-    
+    """
+
     WIRE_TYPE = 2
-    
+
     def dump(self, fp, value):
+        if isinstance(value, six.text_type):
+            value = value.encode("utf-8")
```

Comment:
I think here `.encode` is unwanted since this is the byte string type. Below there's `UnicodeType` which does encoding for a string value.

Reply:
Hm, I looked and couldn't find a better solution here. On Python 2 `str` is the same as `bytes`, and has both `encode` and `decode` methods. On Python 3 `str` isn't the same as `bytes`, but what we want to be storing is something that can go on the wire - so any `str` we use (even as an attribute) must be encoded.

Reply:
The point is, this is a class to serialize a byte string. So a value should go to the wire unchanged. If a user passes a string there, it should lead to an error.